service/history/ndc/activity_state_replicator.go

```diff
@@ -126,64 +130,188 @@ func (r *ActivityStateReplicatorImpl) SyncActivityState(
+		updateMode = persistence.UpdateWorkflowModeBypassCurrent
+	}
+
+	return executionContext.UpdateWorkflowExecutionWithNew(
+		ctx,
+		r.shardContext,
+		updateMode,
+		nil, // no new workflow
+		nil, // no new workflow
+		workflow.TransactionPolicyPassive,
+		nil,
+	)
+}
+
+func (r *ActivityStateReplicatorImpl) SyncActivitiesState(
+	ctx context.Context,
+	request *historyservice.SyncActivitiesRequest,
+) (retError error) {
+	// sync activity info will only be sent from active side, when
+	// 1. activity retry
+	// 2. activity start
+	// 3. activity heart beat
+	// no sync activity task will be sent when active side fail / timeout activity,
+	namespaceID := namespace.ID(request.GetNamespaceId())
+	execution := commonpb.WorkflowExecution{
+		WorkflowId: request.WorkflowId,
+		RunId:      request.RunId,
+	}
+
+	executionContext, release, err := r.workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		r.shardContext,
 		namespaceID,
-		execution.GetWorkflowId(),
-		execution.GetRunId(),
+		execution,
+		workflow.LockPriorityHigh,
+	)
+	if err != nil {
+		// for get workflow execution context, with valid run id
+		// err will not be of type EntityNotExistsError
+		return err
+	}
+	defer func() { release(retError) }()
+
+	mutableState, err := executionContext.LoadMutableState(ctx, r.shardContext)
+	if err != nil {
+		if _, isNotFound := err.(*serviceerror.NotFound); isNotFound {
+			// this can happen if the workflow start event and this sync activity task are out of order
+			// or the target workflow is long gone
+			// the safe solution to this is to throw away the sync activity task
+			// or otherwise, worker attempt will exceed limit and put this message to DLQ
+			return nil
+		}
+		return err
+	}
+	anyEventApplied := false
+	for _, syncActivityInfo := range request.ActivitiesInfo {
+		applied, err := r.syncSingleActivityState(
+			&definition.WorkflowKey{
+				NamespaceID: request.NamespaceId,
+				WorkflowID:  request.WorkflowId,
+				RunID:       request.RunId,
+			},
+			mutableState,
+			syncActivityInfo,
+		)
+		if err != nil {
+			return err
+		}
+		if applied == true {
+			anyEventApplied = true
+		}
+	}
+	if !anyEventApplied {
+		return nil
+	}
+
+	updateMode := persistence.UpdateWorkflowModeUpdateCurrent
+	if state, _ := mutableState.GetWorkflowStateStatus(); state == enumsspb.WORKFLOW_EXECUTION_STATE_ZOMBIE {
+		updateMode = persistence.UpdateWorkflowModeBypassCurrent
+	}
+
+	return executionContext.UpdateWorkflowExecutionWithNew(
+		ctx,
+		r.shardContext,
+		updateMode,
+		nil, // no new workflow
+		nil, // no new workflow
+		workflow.TransactionPolicyPassive,
+		nil,
+	)
+}
+
+func (r *ActivityStateReplicatorImpl) syncSingleActivityState(
+	workflowKey *definition.WorkflowKey,
+	mutableState workflow.MutableState,
+	activitySyncInfo *historyservice.ActivitySyncInfo,
+) (applied bool, retError error) {
+	scheduledEventID := activitySyncInfo.GetScheduledEventId()
+	shouldApply, err := r.testVersionHistory(
```

Comment:
what if 50% of acitvity can be applied and other 50% cannot?
current behavior will just return err?

Reply:
i believe 
1. input validation that all activties belong to the same version history branch
2. before apply any activity, making sure ^ version history branch & current history branch does not diverge
3. split activities into 2 categories
  a. category one: activities with event ID already exists on current version history branch
  b. category two: activities with event ID NOT exists on current version history branch

for 3.b definitely error out with resend error
for 3.a, **can** be applied before return 3.b error

Reply:
> what if 50% of acitvity can be applied and other 50% cannot?
current behavior will just return err?
Yeah, current implementation will return error if any activity failed within the batch. We can improve it if necessary in the future.